cmstestsuite/unit_tests/grading/__init__.py

```diff
@@ -0,0 +1,174 @@
+            "execution_wall_clock_time": execution_wall_clock_time,
+            "execution_memory": execution_memory,
+            "exit_status": exit_status,
+        }
+        if signal is not None:
+            r["signal"] = signal
+        if filename is not None:
+            r["filename"] = filename
+        return r
+
+    def test_success_status_ok(self):
+        self.assertEqual(
+            merge_evaluation_results(
+                self._res(1.0, 2.0, 300, Sandbox.EXIT_OK),
+                self._res(0.1, 0.2, 0.3, Sandbox.EXIT_OK)),
+            self._res(1.1, 2.0, 300.3, Sandbox.EXIT_OK))
+
+    def test_success_first_status_ok(self):
+        self.assertEqual(
+            merge_evaluation_results(
+                self._res(0, 0, 0, Sandbox.EXIT_OK),
+                self._res(0, 0, 0, Sandbox.EXIT_TIMEOUT)),
+            self._res(0, 0, 0, Sandbox.EXIT_TIMEOUT))
+        self.assertEqual(
+            merge_evaluation_results(
+                self._res(0, 0, 0, Sandbox.EXIT_OK),
+                self._res(0, 0, 0, Sandbox.EXIT_FILE_ACCESS, filename="asd")),
+            self._res(0, 0, 0, Sandbox.EXIT_FILE_ACCESS, filename="asd"))
+        self.assertEqual(
+            merge_evaluation_results(
+                self._res(0, 0, 0, Sandbox.EXIT_OK),
+                self._res(0, 0, 0, Sandbox.EXIT_SIGNAL, signal="11")),
+            self._res(0, 0, 0, Sandbox.EXIT_SIGNAL, signal="11"))
+
+    def test_success_first_status_not_ok(self):
+        self.assertEqual(
+            merge_evaluation_results(
+                self._res(0, 0, 0, Sandbox.EXIT_TIMEOUT),
+                self._res(0, 0, 0, Sandbox.EXIT_SIGNAL, signal="11")),
+            self._res(0, 0, 0, Sandbox.EXIT_TIMEOUT))
+        self.assertEqual(
+            merge_evaluation_results(
+                self._res(0, 0, 0, Sandbox.EXIT_FILE_ACCESS, filename="asd"),
+                self._res(0, 0, 0, Sandbox.EXIT_SIGNAL, signal="11")),
+            self._res(0, 0, 0, Sandbox.EXIT_FILE_ACCESS, filename="asd"))
+        self.assertEqual(
+            merge_evaluation_results(
+                self._res(0, 0, 0, Sandbox.EXIT_SIGNAL, signal="9"),
+                self._res(0, 0, 0, Sandbox.EXIT_SIGNAL, signal="11")),
+            self._res(0, 0, 0, Sandbox.EXIT_SIGNAL, signal="9"))
+
+    def test_success_results_are_not_modified(self):
+        r0 = self._res(1.0, 2.0, 300, Sandbox.EXIT_OK)
+        r1 = self._res(0.1, 0.2, 0.3, Sandbox.EXIT_SIGNAL, signal="11")
+        m = merge_evaluation_results(r0, r1)
+        self.assertEqual(
+            m, self._res(1.1, 2.0, 300.3, Sandbox.EXIT_SIGNAL, signal="11"))
+        self.assertEqual(
+            r0, self._res(1.0, 2.0, 300, Sandbox.EXIT_OK))
+        self.assertEqual(
+            r1, self._res(0.1, 0.2, 0.3, Sandbox.EXIT_SIGNAL, signal="11"))
+
+
+class TestWhiteDiff(unittest.TestCase):
+
+    @staticmethod
+    def _diff(s1, s2):
+        return white_diff(StringIO.StringIO(s1), StringIO.StringIO(s2))
```

Comment:
White diff is designed to operate on bytes, not strings.

Reply:
Which also means that all string literals in this test should become `b"..."`. (I guess right now an implicit encoding using utf-8 is happening somewhere)

Reply:
About that, maybe add some tests that use non-ascii-encoded strings...

Reply:
Done.